### PR TITLE
nrf_rpc: Add OT udp api to nRF RPC.

### PR DESCRIPTION
**Pull request by canisLupus1313**
List of APIs added:
    - otUdpOpen
    - otUdpClose
    - otUdpBind
    - otUdpConnect
    - otUdpSend

### CODE REVIEW

**Review by bpienk — APPROVED** [review]

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+		memset(socket, 0, sizeof(nrfUdpSocket));
+	}
+}
+
+static void handle_udp_receive(void *aContext, otMessage *aMessage,
+			       const otMessageInfo *aMessageInfo)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	nrfUdpSocket *socket = (nrfUdpSocket *)aContext;
+	ot_msg_key msg_key = ot_reg_msg_alloc();
```
Comment (SylwesterKonczyk) [inline subsys/net/openthread/rpc/server/ot_rpc_udp_server.c:77]: what if msg_key = 0 ?

File: subsys/net/openthread/rpc/client/ot_rpc_udp_client.c [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+	return error;
+}
+
+otError otUdpBind(otInstance *aInstance, otUdpSocket *aSocket, const otSockAddr *aSockName,
+		  otNetifIdentifier aNetif)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	ot_socket_key key = (ot_socket_key)aSocket;
+	otError error = OT_ERROR_NONE;
+	uint32_t netIf = aNetif;
```
Comment (alxelax) [inline subsys/net/openthread/rpc/client/ot_rpc_udp_client.c:94]:
```suggestion
	uint32_t net_if = aNetif;
```

File: subsys/net/openthread/rpc/client/ot_rpc_udp_client.c [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+static void ot_rpc_cmd_udp_receive_cb(const struct nrf_rpc_group *group,
+				      struct nrf_rpc_cbor_ctx *ctx, void *handler_data)
+{
+	bool decoding_ok;
+	otMessageInfo message_info;
+	ot_msg_key msg_key = 0;
+	ot_socket_key soc_key = 0;
+	otUdpSocket *socket;
+
+	decoding_ok = zcbor_uint_decode(ctx->zs, &soc_key, sizeof(soc_key));
```
Comment (alxelax) [inline subsys/net/openthread/rpc/client/ot_rpc_udp_client.c:144]: I guess it is much optimal to use decoders from `nrf_rpc_serializer.h` and finally `nrf_rpc_decoding_done_and_check`.
However, it is possible to do as part of `Polishing` task.

File: subsys/net/openthread/rpc/common/ot_rpc_common.c [hunk 4]
```diff
@@ -147,3 +149,111 @@ void ot_rpc_report_decoding_error(uint8_t cmd_evt_id)
...
+			       sizeof(aMessageInfo->mPeerPort))) {
+		return false;
+	}
+
+	if (!zcbor_uint_encode(ctx->zs, &aMessageInfo->mHopLimit,
+			       sizeof(aMessageInfo->mHopLimit))) {
+		return false;
+	}
+
+	tmp = aMessageInfo->mEcn;
```
Comment (alxelax) [inline subsys/net/openthread/rpc/common/ot_rpc_common.c:182]: Why do you need `tmp`?

File: subsys/net/openthread/rpc/common/ot_rpc_common.c [hunk 5]
```diff
@@ -147,3 +149,111 @@ void ot_rpc_report_decoding_error(uint8_t cmd_evt_id)
...
+
+	return true;
+}
+
+bool ot_rpc_decode_message_info(struct nrf_rpc_cbor_ctx *ctx, otMessageInfo *aMessageInfo)
+{
+	struct zcbor_string zst;
+	uint8_t tmp;
+
+	if (!zcbor_bstr_decode(ctx->zs, &zst)) {
```
Comment (alxelax) [inline subsys/net/openthread/rpc/common/ot_rpc_common.c:207]: Again, it is more optimal to use nrf rpc deserializer to not check the result of every decoding.

File: subsys/net/openthread/rpc/common/ot_rpc_common.h [hunk 6]
```diff
@@ -10,43 +10,37 @@
...
-	sizeof((dataset)->mDelay) + 1 + \
-	sizeof((dataset)->mPanId) + 1 + \
-	sizeof((dataset)->mChannel) + 1 + \
-	sizeof((dataset)->mPskc.m8) + 1 + \
-	OT_RPC_SECURITY_POLICY_LENGTH((dataset)->mSecurityPolicy) + \
-	sizeof((dataset)->mChannelMask) + 1 + \
-	OT_RPC_OPERATIONAL_COMPONENTS_LENGTH((dataset)->mComponents)
+#include <openthread/ip6.h>
+
+#define OT_RPC_TIMESTAMP_LENGTH(timestamp)                                                         \
```
Comment (alxelax) [inline subsys/net/openthread/rpc/common/ot_rpc_common.h:15]: Could you revert this change? There is no value in this just makes macros representation cumbersome.

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+#include <string.h>
+
+#include "ot_rpc_message_server.h"
+
+#define OT_UDP_SOCKETS 8
+
+typedef struct nrfUdpSocket {
+	ot_socket_key mKey;
+	otUdpSocket mSocket;
+} nrfUdpSocket;
```
Comment (alxelax) [inline subsys/net/openthread/rpc/server/ot_rpc_udp_server.c:30]: Again incompatible style.
Why do you need separate `type`?

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+
+#include <openthread/udp.h>
+
+#include <zephyr/net/openthread.h>
+#include <zcbor_decode.h>
+#include <string.h>
+
+#include "ot_rpc_message_server.h"
+
+#define OT_UDP_SOCKETS 8
```
Comment (alxelax) [inline subsys/net/openthread/rpc/server/ot_rpc_udp_server.c:25]: Why 8? Could it be configurable?

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+
+typedef struct nrfUdpSocket {
+	ot_socket_key mKey;
+	otUdpSocket mSocket;
+} nrfUdpSocket;
+
+static nrfUdpSocket sSockets[OT_UDP_SOCKETS];
```
Comment (alxelax) [inline subsys/net/openthread/rpc/server/ot_rpc_udp_server.c:32]:
```suggestion
static nrf_udp_socket sockets[OT_UDP_SOCKETS];
```

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+	nrfUdpSocket *ret = nrf_udp_find_socket(key);
+
+	if (ret != NULL) {
+		return ret;
+	}
+
+	ret = nrf_udp_find_socket(0);
+
+	if (ret != NULL) {
+		memset(ret, 0, sizeof(nrfUdpSocket));
```
Comment (alxelax) [inline subsys/net/openthread/rpc/server/ot_rpc_udp_server.c:56]: Why is it required? It should be zero if it is empty. Otherwise, searching by 0 doesn't have sense.

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+static void nrf_udp_free_socket(ot_socket_key key)
+{
+	nrfUdpSocket *socket = nrf_udp_find_socket(key);
+
+	if (socket != NULL) {
+		memset(socket, 0, sizeof(nrfUdpSocket));
+	}
+}
+
+static void handle_udp_receive(void *aContext, otMessage *aMessage,
```
Comment (alxelax) [inline subsys/net/openthread/rpc/server/ot_rpc_udp_server.c:72]:
```suggestion
static void handle_udp_receive(void *context, otMessage *message,
```

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+			       const otMessageInfo *aMessageInfo)
```
Comment (alxelax) [inline subsys/net/openthread/rpc/server/ot_rpc_udp_server.c:73]:
```suggestion
			       const otMessageInfo *message_info)
```

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+	if (!ot_rpc_encode_message_info(&ctx, aMessageInfo)) {
+		NRF_RPC_CBOR_DISCARD(&ot_group, ctx);
+		goto exit;
+	}
+
+	nrf_rpc_cbor_cmd_rsp_no_err(&ot_group, OT_RPC_CMD_UDP_RECEIVE_CB, &ctx);
+
+	nrf_rpc_cbor_decoding_done(&ot_group, &ctx);
+
+	ot_msg_free(msg_key);
```
Comment (alxelax) [inline subsys/net/openthread/rpc/server/ot_rpc_udp_server.c:104]: In case of encoding error, there will be memory leakage.

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c [hunk 14]
```diff
@@ -0,0 +1,390 @@
...
+static void ot_rpc_udp_send(const struct nrf_rpc_group *group, struct nrf_rpc_cbor_ctx *ctx,
+			    void *handler_data)
+{
+	struct nrf_rpc_cbor_ctx rsp_ctx;
+	otError error = OT_ERROR_NONE;
+	bool decoded_ok;
+	ot_socket_key soc_key = 0;
+	ot_msg_key msg_key = 0;
+	nrfUdpSocket *socket;
+	otMessageInfo messageInfo;
```
Comment (alxelax) [inline subsys/net/openthread/rpc/server/ot_rpc_udp_server.c:158]:
```suggestion
	otMessageInfo message_info;
```

**Review by alxelax — COMMENTED** [review]
I do not want to block PR but seems it requires significant cleaning up.

File: subsys/net/openthread/rpc/common/ot_rpc_types.h [hunk 15]
```diff
@@ -28,3 +31,6 @@ enum ot_rpc_link_mode_offsets {
 };
 
 typedef uint32_t ot_msg_key;
+typedef uint32_t ot_socket_key;
```
Comment (SylwesterKonczyk) [inline subsys/net/openthread/rpc/common/ot_rpc_types.h:34]: ot_msg_key, ot_socket_key - Would it be beneficial to document usage scenarios of above types?

File: subsys/net/openthread/rpc/server/ot_rpc_message_server.h [hunk 16]
```diff
@@ -14,5 +14,6 @@
 ot_msg_key ot_reg_msg_alloc(void);
 void ot_msg_free(ot_msg_key key);
 otMessage *ot_msg_get(ot_msg_key key);
+bool ot_msg_set(ot_msg_key key, otMessage *msg);
```
Comment (Damian-Nordic) [inline subsys/net/openthread/rpc/server/ot_rpc_message_server.h:17]: `ot_reg_msg_alloc` could take `otMessage*` as a parameter and then this wouldn't be needed.

File: subsys/net/openthread/rpc/common/ot_rpc_common.c (hunk 4 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/common/ot_rpc_common.c:182]: `mEcn` is a bit field to which I can't get address to.

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c (hunk 7 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/server/ot_rpc_udp_server.c:30]: Coding style changed.

As for the new type. on this site we need registry with preallocated sockets. Unlike ot message otUdpSocket is also a element in list (ptr to next is embedded into this  data type). This means that I have to leave all fields intact. As for the key it's actually a pointer to same structure on client site. Instead of creating similar registry I'm keeping the e callback and context data in this struct.

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c (hunk 1 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/server/ot_rpc_udp_server.c]: fixed.

File: samples/nrf_rpc/protocols_serialization/client/src/ot_shell.c [hunk 17]
```diff
@@ -16,6 +16,10 @@
 
 #include <string.h>
 
+static otUdpSocket sUdpSocket;
+static const char UDP_PAYLOAD[] = "Hello OpenThread World!";
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply samples/nrf_rpc/protocols_serialization/client/src/ot_shell.c]: fixed.

File: samples/nrf_rpc/protocols_serialization/client/src/ot_shell.c [hunk 18]
```diff
@@ -454,6 +457,82 @@ static int cmd_test_message(const struct shell *sh, size_t argc, char *argv[])
 	return 0;
 }
 
+static void handleUdpReceive(void *aContext, otMessage *aMessage, const otMessageInfo *aMessageInfo)
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply samples/nrf_rpc/protocols_serialization/client/src/ot_shell.c]: fixed.

File: samples/nrf_rpc/protocols_serialization/client/src/ot_shell.c [hunk 19]
```diff
@@ -454,6 +457,82 @@ static int cmd_test_message(const struct shell *sh, size_t argc, char *argv[])
...
+	if (read > 0) {
+		printk("RECEIVED: '%s'\n", buf);
+	} else {
+		printk("message empty\n");
+	}
+}
+
+static int cmd_test_udp_init(const struct shell *sh, size_t argc, char *argv[])
+{
+	otSockAddr listenSockAddr;
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply samples/nrf_rpc/protocols_serialization/client/src/ot_shell.c]: fixed.

File: samples/nrf_rpc/protocols_serialization/client/src/ot_shell.c [hunk 20]
```diff
@@ -16,6 +16,10 @@
 
 #include <string.h>
 
+static otUdpSocket sUdpSocket;
+static const char UDP_PAYLOAD[] = "Hello OpenThread World!";
+static const uint16_t port = 1212;
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply samples/nrf_rpc/protocols_serialization/client/src/ot_shell.c]: fixed.

File: samples/nrf_rpc/protocols_serialization/client/src/ot_shell.c [hunk 21]
```diff
@@ -454,6 +457,82 @@ static int cmd_test_message(const struct shell *sh, size_t argc, char *argv[])
...
+	otUdpBind(NULL, &sUdpSocket, &listenSockAddr, OT_NETIF_THREAD);
+
+	return 0;
+}
+
+static int cmd_test_udp_send(const struct shell *sh, size_t argc, char *argv[])
+{
+	otError error = OT_ERROR_NONE;
+	otMessage *message = NULL;
+	otMessageInfo messageInfo;
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply samples/nrf_rpc/protocols_serialization/client/src/ot_shell.c]: fixed.

File: samples/nrf_rpc/protocols_serialization/client/src/ot_shell.c [hunk 22]
```diff
@@ -454,6 +457,82 @@ static int cmd_test_message(const struct shell *sh, size_t argc, char *argv[])
...
+exit:
+	if (error != OT_ERROR_NONE && message != NULL) {
+		otMessageFree(message);
+	}
+	return 0;
+}
+
+static int cmd_test_udp_close(const struct shell *sh, size_t argc, char *argv[])
+{
+	otUdpClose(NULL, &sUdpSocket);
```
[the remark this answers is not in the document]
Comment (canisLupus1313) [reply samples/nrf_rpc/protocols_serialization/client/src/ot_shell.c]: fixed.

File: subsys/net/openthread/rpc/client/ot_rpc_udp_client.c (hunk 2 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/client/ot_rpc_udp_client.c]: fixed.

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c (hunk 9 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/server/ot_rpc_udp_server.c]: fixed.

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c (hunk 8 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/server/ot_rpc_udp_server.c:25]: it could :)

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c (hunk 10 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/server/ot_rpc_udp_server.c]: Yo are correct removed the code.

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c (hunk 12 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/server/ot_rpc_udp_server.c]: fixed.

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c (hunk 14 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/server/ot_rpc_udp_server.c]: fixed.

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c (hunk 13 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/server/ot_rpc_udp_server.c]: fixed.

File: subsys/net/openthread/rpc/server/ot_rpc_message_server.h (hunk 16 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/server/ot_rpc_message_server.h]: fixed.

File: subsys/net/openthread/rpc/server/ot_rpc_udp_server.c (hunk 11 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/server/ot_rpc_udp_server.c]: fixed.

File: subsys/net/openthread/rpc/common/ot_rpc_common.h (hunk 6 above)
Comment (canisLupus1313) [reply subsys/net/openthread/rpc/common/ot_rpc_common.h]: fixed.

**Review by SylwesterKonczyk — APPROVED** [review]
